### PR TITLE
U2F keyladder support

### PR DESCRIPTION
**Pull request by phil-levis-google**
This PR adds limited support for the H1B keyladder and has the U2F application invoke the keyladder to generate its keying material. This involves some changes to the digest interface to allow HMAC digests involving keyladder "certificates". The system call API is very narrow -- it allows a process invoke a step of the keyladder for a particular cert.

**Remember to run `make prtest` and paste the output here (replace this line)**
```
Bldr |53803887
retry|0
oops?|no
Himg =8D1E52A5..4B60EE60
Hfss =635D6FFB..6E85CEE3
Hinf =D0FC7426..BC208A3A
ump @44000
ock: booted in 68182 tics; initializing USB and loading processes.
Tock: starting main loop.
 
Startingtests.
Running test basic_test
Cat video count: 9001
What we eat: cafef00d
inished test basic_test. Result: ucceeded.
TEST_FINISHED: SUCCESS
```

### CODE REVIEW

File: golf2/src/digest.rs [hunk 1]
```diff
@@ -53,9 +53,11 @@ impl<'a, E: DigestEngine + 'a> DigestDriver<'a, E> {
 
 impl<'a, E: DigestEngine> Driver for DigestDriver<'a, E> {
     fn command(&self, minor_num: usize, r2: usize, _r3: usize, caller_id: AppId) -> ReturnCode {
+        //print!("DigestEngine: command({}, {}, {})\n", minor_num, r2, _r3);
```
Comment (jrvanwhy) [inline golf2/src/digest.rs]: Line 56 appears to be a leftover debugging statement -- please remove.

File: golf2/src/digest.rs [hunk 2]
```diff
@@ -113,39 +116,74 @@ impl<'a, E: DigestEngine> Driver for DigestDriver<'a, E> {
...
             },
             // Finalize hash and output to output buffer (arg: unused)
-            2 => {
+            3 => {
                 self.apps
                     .enter(caller_id, |app_data, _| {
                         match self.current_user.get() {
                             Some(cur) if cur.idx() == caller_id.idx() => {}
                             _ => {
+                                //debug!("Finalize busy, {}\n", caller_id.idx());
```
Comment (jrvanwhy) [inline golf2/src/digest.rs]: Another debugging statement that probably should be removed.

File: golf2/src/digest.rs [hunk 3]
```diff
@@ -113,39 +116,74 @@ impl<'a, E: DigestEngine> Driver for DigestDriver<'a, E> {
...
                     })
                     .unwrap_or(ReturnCode::ENOMEM)
             },
+            4 => {
+                ReturnCode::SUCCESS
+                    /*if self.current_user.get().is_some() {
+                    ReturnCode::EBUSY
+                } else {
+                    ReturnCode::SUCCESS
+                }*/
```
Comment (jrvanwhy) [inline golf2/src/digest.rs]: Leftover code?

File: h1b/src/lib.rs [hunk 4]
```diff
@@ -128,4 +128,6 @@ pub unsafe fn init() {
     cortexm3::nvic::disable_all();
     cortexm3::nvic::clear_all_pending();
     cortexm3::nvic::enable_all();
+
+    //cortexm3::nvic::Nvic::new(110).disable(); // SHA spins
```
Comment (jrvanwhy) [inline h1b/src/lib.rs]: Revert?

File: userspace/libgolf2/digest_syscalls.c [hunk 5]
```diff
@@ -36,6 +51,11 @@ int tock_digest_hash_initialize(TockDigestMode mode) {
   return rval;
 }
 
+int tock_digest_cert_initialize(uint32_t cert) {
+  int rval = command(H1B_DRIVER_DIGEST, TOCK_DIGEST_CMD_CERT_INIT, cert, 0);
```
Comment (jrvanwhy) [inline userspace/libgolf2/digest_syscalls.c]: This can be a 1-liner, correct?

File: userspace/u2f_app/u2f_transport.c [hunk 6]
```diff
@@ -407,13 +409,13 @@ static void u2fhid_cmd_init(U2FHID_FRAME *f_p) {
...
-          response.init.data[1], response.init.data[2], response.init.data[3],
-          response.init.data[4], response.init.data[5], response.init.data[6],
-          response.init.data[7]);
+  //printf("Response Frame -> cid:%08lx cmd:%02x ", response.cid,
+  //        response.init.cmd);
+  //printf("bcnth:%02x bcntl:%02x ", response.init.bcnth, response.init.bcntl);
+  //printf("data:%0X%0X%0X%0X%0X%0X%0X%0X\n\n", response.init.data[0],
+  //        response.init.data[1], response.init.data[2], response.init.data[3],
+  //        response.init.data[4], response.init.data[5], response.init.data[6],
+  //        response.init.data[7]);
```
Comment (jrvanwhy) [inline userspace/u2f_app/u2f_transport.c]: I suspect these printf statements should be removed.

If we still want them, we should probably define a "debug_printf" or similar that is a no-op by default and can be enabled through a debug option (`make devprogram`?).

**Review by jrvanwhy — CHANGES_REQUESTED** [review]
In general, we should probably have documentation for each system call driver. A .md file in docs/ with a table identifying the behavior of each minor_num would suffice.

File: userspace/u2f_app/u2f_transport.c (hunk 6 above)
Comment (phil-levis-google) [reply userspace/u2f_app/u2f_transport.c]: I agree that commented out printf statements are bad.

The issue I'm wrestling with is that when something goes wrong, you often want information from many different layers at the same time, but within a particular code path. I.e., you want debugging information about the control channel (EP0) across different parts of the stack, or you want data path information (EP1) across different parts of the stack. You can just reintroduce debug output each time, but this is a pain. So in the best case we have a set of debug flags you can turn on/off. You can see this in the three different debug macros within usb/mod.rs.

It's worth noting this particular printf() is from the original C code. :)

File: golf2/src/digest.rs (hunk 1 above)
Comment (phil-levis-google) [reply golf2/src/digest.rs]: Done.

File: golf2/src/digest.rs (hunk 2 above)
Comment (phil-levis-google) [reply golf2/src/digest.rs]: Done.

Comment (phil-levis-google) [reply golf2/src/digest.rs]: Should it be in doc/ or in libgolf2/ ?

Comment (jrvanwhy) [reply golf2/src/digest.rs]: I don't have a strong opinion on the topic. Keeping everything in doc/ makes it much more findable, with the risk that someone working with code forgets that the documentation exists. If you're putting documentation into libgolf2/ then it probably makes sense for that documentation to be in the form of a comment in the code.

**Review by jrvanwhy — APPROVED** [review]

**Review by jrvanwhy — APPROVED** [review]

File: userspace/u2f_app/u2f_transport.c (hunk 6 above)
Comment (jrvanwhy) [reply userspace/u2f_app/u2f_transport.c]: Yeah, this is a more general problem that we should solve at some point. A better debugging system would let us leave the debug messages in the code but make them no-ops (that compile away entirely) by default.

File: golf2/src/digest.rs (hunk 2 above)
Comment (phil-levis-google) [reply golf2/src/digest.rs]: Put it in doc/.

File: h1b/src/lib.rs (hunk 4 above)
Comment (phil-levis-google) [reply h1b/src/lib.rs]: Removed.

File: golf2/src/digest.rs (hunk 3 above)
Comment (phil-levis-google) [reply golf2/src/digest.rs]: Ah, should have been commented back in; currently not used but needed in API.

File: userspace/libgolf2/digest_syscalls.c (hunk 5 above)
Comment (phil-levis-google) [reply userspace/libgolf2/digest_syscalls.c]: Yes. Previously had debug output. Made into a one-liner.

File: userspace/u2f_app/u2f_transport.c (hunk 6 above)
Comment (phil-levis-google) [reply userspace/u2f_app/u2f_transport.c]: Just cut it.

**Pull request merged**